bench/hmap.ml

```diff
@@ -0,0 +1,147 @@
+module type Interface = sig
+  type 'a t
+  type key = int
+
+  val create : size_exponent:int -> 'a t
+  val add : key -> 'a -> 'a t -> bool
+  val find_opt : key -> 'a t -> 'a option
+  val remove : key -> 'a t -> bool
+end
+
+module Locked : Interface = struct
+  type key = int
+  type 'a t = (key, 'a) Hashtbl.t
+
+  let mutex = Mutex.create ()
+
+  let with_mutex f =
+    Mutex.lock mutex;
+    let v = f () in
+    Mutex.unlock mutex;
+    v
+
+  let create ~size_exponent =
+    let size = Int.shift_left 1 size_exponent in
+    Hashtbl.create size
```

Comment:
Note that this will use the generic hash which is very slow.

Reply:
agreed, if  you compare with `Hashtbl` you absolutely have to use `Hashtbl.Make`, for example with this:

```
type t = int
let equal : t -> t -> bool = (=)  (* type annotation might be important to avoid polymorphic compare *)
let hash = <your hash here>
```

Reply:
I would recommend using `Fun.id` as the hash function as I mentioned in a previous comment/question. The goal being to measure performance of the hash table intenals rather than the hash function.

Reply:
I agree that the hash shouldn't be too expensive, *BUT* a bad hash function (such as `Fun.id`) will also impact the hashtable's performance, by potentially making some buckets a lot larger than others.

In containers I use a reimplementation of FNV ([godbolt here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:ocaml,selection:(endColumn:25,endLineNumber:18,positionColumn:25,positionLineNumber:18,selectionStartColumn:25,selectionStartLineNumber:18,startColumn:25,startLineNumber:18),source:'let+hash+(n+:+int)+:+int+%3D%0A++let+offset_basis+%3D+0xcbf29ce484222325L+in%0A++let+prime+%3D+0x100000001b3L+in%0A%0A++let+h+%3D+ref+offset_basis+in%0A++for+k+%3D+0+to+7+do%0A++++(h+:%3D+Int64.(mul+!!h+prime))%3B%0A++++(*+h+:%3D+h+xor+(k-th+byte+of+n)+*)%0A++++h+:%3D+Int64.(logxor+!!h+(of_int+((n+lsr+(k+*+8))+land+0xff)))%0A++done%3B%0A++(*+truncate+back+to+int+and+remove+sign+*)%0A++Int64.to_int+!!h+land+max_int%0A%0A%0Alet+()+%3D%0A++let+i+%3D+Sys.opaque_identity+42+in%0A++let+h+%3D+hash+i+in%0A++Printf.printf+%22%25d%5Cn%22+h'),l:'5',n:'0',o:'OCaml+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:ocaml5000flambda,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:ocaml,libs:!(),options:'-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+ocamlopt+5.0.0-flambda+(Editor+%231)',t:'0')),k:50,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)) which should not be _too_ expensive.

Reply:
Note that the benchmark uses random integers as keys.  More generally it is not difficult to avoid bad key sequences in benchmarks.  At a minimum, one should use the same user specified hash for all tables.